server/src/delta_cache.rs

```diff
@@ -0,0 +1,286 @@
+            max_length,
+            events: VecDeque::new(),
+            hydration_event: hydration_event.clone(),
+        };
+        cache.add_base_event_from_hydration(hydration_event);
+        cache
+    }
+
+    fn add_base_event_from_hydration(&mut self, hydration_event: DeltaHydrationEvent) {
+        if let Some(last_feature) = hydration_event.features.last().cloned() {
+            self.add_events(vec![DeltaEvent::FeatureUpdated {
+                event_id: hydration_event.event_id,
+                feature: last_feature,
+            }]);
+        }
+    }
+
+    pub fn add_events(&mut self, events: Vec<DeltaEvent>) {
+        for event in events.into_iter() {
+            self.events.push_back(event.clone());
+            self.update_hydration_event(event);
+            if self.events.len() > self.max_length {
+                self.events.pop_front();
+            }
+        }
+    }
+
+    pub fn get_events(&self) -> &VecDeque<DeltaEvent> {
+        &self.events
+    }
+
+    pub fn is_missing_revision(&self, revision_id: u32) -> bool {
+        !self.events.iter().any(|event| event.get_event_id() == revision_id as i32)
+    }
+
+    pub fn get_hydration_event(&self) -> &DeltaHydrationEvent {
+        &self.hydration_event
+    }
+
+    fn update_hydration_event(&mut self, event: DeltaEvent) {
+        match event {
+            DeltaEvent::FeatureUpdated { event_id, feature } => {
+                if let Some(existing) = self.hydration_event.features.iter_mut().find(|f| f.name == feature.name) {
+                    *existing = feature;
+                } else {
+                    self.hydration_event.features.push(feature);
+                }
+                self.hydration_event.event_id = event_id;
+            }
+            DeltaEvent::FeatureRemoved { event_id, feature_name } => {
+                self.hydration_event.features.retain(|f| f.name != feature_name);
+                self.hydration_event.event_id = event_id;
+            }
+            DeltaEvent::SegmentUpdated { event_id, segment } => {
+                if let Some(existing) = self.hydration_event.segments.iter_mut().find(|s| s.id == segment.id) {
+                    *existing = segment;
+                } else {
+                    self.hydration_event.segments.push(segment);
+                }
+                self.hydration_event.event_id = event_id;
+            }
+            DeltaEvent::SegmentRemoved { event_id, segment_id } => {
+                self.hydration_event.segments.retain(|s| s.id != segment_id);
+                self.hydration_event.event_id = event_id;
+            }
+            DeltaEvent::Hydration { .. } => {
+               // do nothing, as hydration will never end up in update events
+            }
+        }
+    }
+}
+
+
+#[test]
+fn test_update_hydration_event_and_remove_event_when_over_limit() {
+    let base_event = DeltaHydrationEvent {
+        event_id: 1,
+        features: vec![
+            ClientFeature {
```

Comment:
IIRC correctly `ClientFeature` derives `Default`, which means these big ass declarations of ClientFeature can be simplified to something like

`ClientFeature {
  name: "test-flag".to_string(),
  ///other stuff you care about
  ..ClientFeatures::default()}
}`

Makes the tests much easier to navigate and saves future maintenance work when someone adds a new field to `ClientFeature`